### PR TITLE
first version of CSR attribute as template

### PR DESCRIPTION
**Pull request by mcr**
For discussion on Tuesday.

### CODE REVIEW

**Comment by DDvO** [comment]
Well, we can reference {{RFC8295, Appendix B}},
but as I had already commented on this approach by email on 2023-07-28 09:33 CEST:

The general idea is quite the same as the suggestion I made on June 20th. (BTW, as mentioned before, essentially the same idea is used also for CMP in
https://datatracker.ietf.org/doc/html/draft-ietf-lamps-lightweight-cmp-profile#name-get-certificate-request-tem.)

Yet the solution given in RFC 8295 differs from our new proposal in several respects, which include:

1.  It allows giving a sequence of CSR templates, which presumably is not needed for the EST context (and at least so far not intended to support).

2. It also provides the possibility to give a CertReqMsg (in CRMF format), which is not needed in the EST context as it uses CSRs in PKCS#10 format only.
3. For both variants, it requires giving a full CSR structure, which in the PKCS#10 case implies mandatory inclusion of a public key and a self-signature.
4. For the PKCS#10 variant, the type is TaggedCertificationRequest (see definition in https://www.rfc-editor.org/rfc/rfc5272#section-3.2.1.2.1),
    which includes a bodyPartID field not needed for the EST use case.

To me, point 3. makes the approach sketched in RFC 8295 Appendix B not really feasible or at least needlessly ugly,
as the server side would have to use a fake/temporary key pair just to satisfy the structural requirements of PKCS#10.

Note that our new proposal does not include a full (PKCS#10) CSR but just the "to-be-signed" part:
```
CertificationRequestInfo ::= SEQUENCE {
        version       INTEGER { v1(0) } (v1,...),
        subject       Name,
        subjectPKInfo SubjectPublicKeyInfo{{ PKInfoAlgorithms }},
        attributes    [0] Attributes{{ CRIAttributes }}
   }
```
This way, we avoid the need to provide the (for the given use case) useless and awkward self-signature.
Moreover, as not mentioned in detail in yesterday's talk, we also foresee the options to

* either specify in the AlgorithmIdentifier of the subjectPKInfo field the type of the key pair the client side is requested use
 * or to give a NULL OID in this AlgorithmIdentifier to indicate that the server has no requirements on the key type.

Sean wrote:
> Note that the appendix is informative and maybe not as detailed as it might have been.  I support the idea of moving the idea to a normative section of a document!

Pleased to hear!
Indeed, our new proposal for the upcoming improvement of the CsrAttrs aspect of RFC 7030 is going to provide all detail needed for practical use -
in particular, a specification of how the server shall specify that certain (sub-)values of certain CSR fields should be filled in by the client,
basically as done already in https://datatracker.ietf.org/doc/html/draft-ietf-lamps-lightweight-cmp-profile#name-get-certificate-request-tem

File: lamps-rfc7030-csrattrs.mkd [hunk 1]
```diff
@@ -153,6 +153,29 @@ MUST NOT include more than one element with a particiular extnID.
...
+Instead, a CSR object is returned with various fields filled out, and other fields waiting to be filled in.
+This is also the method suggested by the CMP XYZ MECHANISM.
+
+~~~~
+    CertificationRequest ::= SEQUENCE {
+        certificationRequestInfo CertificationRequestInfo,
+        signatureAlgorithm AlgorithmIdentifier{{ SignatureAlgorithms }},
+        signature          BIT STRING
+    }
+    CertificationRequestInfo ::= SEQUENCE {
```
Comment (mcr) [inline lamps-rfc7030-csrattrs.mkd]:
```suggestion
    CertificationRequestInfo ::= SEQUENCE {
```

File: lamps-rfc7030-csrattrs.mkd [hunk 2]
```diff
@@ -153,6 +153,24 @@ MUST NOT include more than one element with a particiular extnID.
 With this understanding, the needs of {{RFC8994}} and {{RFC8995}} are satisfied
 with no change to the bits on the wire.
 
+## Alternative: Use of CSR templates
+
+{{RFC8295, Appendix B}} suggests an alternative for the piecemeal specification of attributes that {{RFC7030}} documented.
+Instead, a CSR object is returned with various fields filled out, and other fields waiting to be filled in.
+This is also the method suggested by the CMP XYZ MECHANISM.
```
Comment (mcr) [inline lamps-rfc7030-csrattrs.mkd]:
```suggestion
This is also the method defined CMP Update {{[I-D.ietf-lamps-cmp-updates}} and the Lightweight CMP profile {{I-D.ietf-lamps-lightweight-cmp-profile, Section 4.3.3}}, which uses a CSR template in CRMF.
```

File: lamps-rfc7030-csrattrs.mkd [hunk 3]
```diff
@@ -153,6 +153,24 @@ MUST NOT include more than one element with a particiular extnID.
...
+~~~~
+    CertificationRequestInfo ::= SEQUENCE {
+        version       INTEGER { v1(0) } (v1,...),
+        subject       Name,
+        subjectPKInfo SubjectPublicKeyInfo{{ PKInfoAlgorithms }},
+        attributes    [0] Attributes{{ CRIAttributes }}
+        }
+~~~~
+
+The entire CSR is included as a new Attribute valid with OID TBD.
```
Comment (mcr) [inline lamps-rfc7030-csrattrs.mkd]:
```suggestion
The CertificationRequestInfo is included as a new Attribute valid with OID TBD.  This avoids a useless fake signature in the CSR object.
The version code is always v1 (0). As shown in the example below, any empty values in the subject DN, and in any included X509v3 extensions are expected to be filled in by the client.

The SubjectPublicKeyInfo field must be present, but it MUST have an empty bit string for the key, as the server does not know what key will be used.  The server MAY specify (in the OID), the type of the key to use, but otherwise the OID type MUST be NULL.
```

File: lamps-rfc7030-csrattrs.mkd [hunk 4]
```diff
@@ -153,6 +153,27 @@ MUST NOT include more than one element with a particiular extnID.
...
+~~~~
+    CertificationRequestInfo ::= SEQUENCE {
+        version       INTEGER { v1(0) } (v1,...),
+        subject       Name,
+        subjectPKInfo SubjectPublicKeyInfo{{ PKInfoAlgorithms }},
+        attributes    [0] Attributes{{ CRIAttributes }}
+        }
+~~~~
+
+The CertificationRequestInfo is included as a new Attribute valid with OID TBD.  This avoids a useless fake signature in the CSR object.
```
Comment (mcr) [inline lamps-rfc7030-csrattrs.mkd]:
```suggestion
The CertificationRequestInfo is included as a new Attribute in the CsrAttrs elements sequence.
 It has the type OID TBD.  This attribute MUST be the only element.  
 
 The use of the CertificationRequestInfo avoids a useless fake signature in the CSR object.
```

File: lamps-rfc7030-csrattrs.mkd [hunk 5]
```diff
@@ -153,6 +153,30 @@ MUST NOT include more than one element with a particiular extnID.
...
+    CertificationRequestInfo ::= SEQUENCE {
+        version       INTEGER { v1(0) } (v1,...),
+        subject       Name,
+        subjectPKInfo SubjectPublicKeyInfo{{ PKInfoAlgorithms }},
+        attributes    [0] Attributes{{ CRIAttributes }}
+        }
+~~~~
+
+The CertificationRequestInfo is included as a new Attribute in the CsrAttrs elements sequence.
+ It has the type OID TBD.  This attribute MUST be the only element.
```
Comment (mcr) [inline lamps-rfc7030-csrattrs.mkd]:
```suggestion
 It has the type OID TBD.  This attribute MUST be the only element.  
 Legacy servers MAY continue to use the {{RFC7030}} style piecemeal attribute/value pairs, as long as this new attribute is not also used.
```

File: lamps-rfc7030-csrattrs.mkd [hunk 6]
```diff
@@ -153,6 +153,32 @@ MUST NOT include more than one element with a particiular extnID.
...
+        subject       Name,
+        subjectPKInfo SubjectPublicKeyInfo{{ PKInfoAlgorithms }},
+        attributes    [0] Attributes{{ CRIAttributes }}
+        }
+~~~~
+
+The CertificationRequestInfo is included as a new Attribute in the CsrAttrs elements sequence.
+ It has the type OID TBD.  This attribute MUST be the only element.  
+ Legacy servers MAY continue to use the {{RFC7030}} style piecemeal attribute/value pairs, as long as this new attribute is not also used.
+ Legacy servers MAY continue to use the {{RFC7030}} style piecemeal attribute/value pairs, as long as this new attribute is not also used.
```
Comment (mcr) [inline lamps-rfc7030-csrattrs.mkd]:
```suggestion
 Legacy servers MAY continue to use the {{RFC7030}} style piecemeal attribute/value pairs, and MAY also include the template style described here.
 Clients which understand both MUST use the template only, and ignore all other CSRattrs elements.  Older clients will ignore this new element.
 
```

File: lamps-rfc7030-csrattrs.mkd [hunk 7]
```diff
@@ -153,6 +153,33 @@ MUST NOT include more than one element with a particiular extnID.
...
+    CertificationRequestInfo ::= SEQUENCE {
+        version       INTEGER { v1(0) } (v1,...),
+        subject       Name,
+        subjectPKInfo SubjectPublicKeyInfo{{ PKInfoAlgorithms }},
+        attributes    [0] Attributes{{ CRIAttributes }}
+        }
+~~~~
+
+The CertificationRequestInfo is included as a new Attribute in the CsrAttrs elements sequence.
+ It has the type OID TBD.  This attribute MUST be the only element.
```
Comment (mcr) [inline lamps-rfc7030-csrattrs.mkd]:
```suggestion
 It has the type OID TBD.  
```
